### PR TITLE
Cosmos DB - Add PatchItem support

### PR DESCRIPTION
**Pull request by taiyokato**
Hi, first time contributing.

This PR addresses https://github.com/Azure/azure-sdk-for-go/issues/18682

Implemented PatchItem method based on existing PUT/POST methods.
This also includes Patch{Add, Replace, Set, Increment}Item wrapper methods of PatchItem per requested in the issue.

The choice of theses wrapper methods are based on the [Java SDK reference](https://learn.microsoft.com/en-us/java/api/com.azure.cosmos.models.cosmospatchoperations?view=azure-java-stable#method-summary)

Tested PatchItem in my personal project with CosmosDB emulator and worked fine. Other wrapper methods are calling PatchItem internally so should work as well.

Sorry, unsure how to version this in azcosmos/CHANGELOG.md

- [x] The purpose of this PR is explained in this or a referenced issue.
- [x] The PR does not update generated files.
   - These files are managed by the codegen framework at [Azure/autorest.go][].
- [x] Tests are included and/or updated for code changes.
- [x] Updates to [CHANGELOG.md][] are included.
- [x] MIT license headers are included in each file.

[Azure/autorest.go]: https://github.com/Azure/autorest.go
[CHANGELOG.md]: https://github.com/Azure/azure-sdk-for-go/blob/main/CHANGELOG.md

### CODE REVIEW

**Comment by taiyokato** [comment]
Forgot remove op so added that

File: sdk/data/azcosmos/cosmos_container.go
Comment (ealsur) [inline sdk/data/azcosmos/cosmos_container.go:454]: Having a single method that receives a list of Patch Operations would reduce the need for all these other methods

File: sdk/data/azcosmos/go.mod
Comment (ealsur) [inline sdk/data/azcosmos/go.mod:6]: What was the reason for this increment?

**Review by ealsur — COMMENTED** [review]
Thanks for your valuable contribution! The main point that I see is that the public surface needs a bit of work (compared to other languages) to avoid accidents on getting the wire contract wrong.

File: sdk/data/azcosmos/cosmos_container.go
Comment (taiyokato) [reply sdk/data/azcosmos/cosmos_container.go:372]: Yes, I was thinking of making something similar but instead decided to leave it to user side.
Will try making some contract structs like java does.

File: sdk/data/azcosmos/go.mod
Comment (taiyokato) [reply sdk/data/azcosmos/go.mod:6]: Don't know what happened here. Maybe unnecessary but go mod tidy did something there so thought it might be needed. Might delete it later

File: sdk/data/azcosmos/cosmos_container.go
Comment (taiyokato) [reply sdk/data/azcosmos/cosmos_container.go:454]: Will keep working on this

File: sdk/data/azcosmos/cosmos_container.go
Comment (ealsur) [reply sdk/data/azcosmos/cosmos_container.go:372]: While giving the user the freedom makes the SDK simpler, it also opens the door for someone spending hours debugging an issue because they simply had a typo on the JSON contract. This will also reduce the number of methods in the container to just 1.

File: sdk/data/azcosmos/go.mod
Comment (ealsur) [reply sdk/data/azcosmos/go.mod:6]: @jhendrixMSFT do you know if this side effect of `go mod tidy` is expected?

File: sdk/data/azcosmos/README.md
Comment (ealsur) [inline sdk/data/azcosmos/README.md:146]: Please also update the CHANGELOG. This would go into the "Features Added" section of the 0.3.3 (Unreleased). With some text like "* Adds Patch support*

File: sdk/data/azcosmos/cosmos_container.go
Comment (ealsur) [inline sdk/data/azcosmos/cosmos_container.go:368]: Once we land on the final public API, adding one example to https://github.com/Azure/azure-sdk-for-go/blob/main/sdk/data/azcosmos/example_test.go named `ExampleContainerClient_PatchItem` would be great as it would appear in the go pkg docs

File: sdk/data/azcosmos/go.mod
Comment (jhendrixMSFT) [reply sdk/data/azcosmos/go.mod:6]: Unfortunately yes.  The net change is no impact though.

File: sdk/data/azcosmos/README.md
Comment (taiyokato) [reply sdk/data/azcosmos/README.md:146]: Done

File: sdk/data/azcosmos/cosmos_container.go
Comment (taiyokato) [reply sdk/data/azcosmos/cosmos_container.go:367]: Done

**Comment by taiyokato** [comment]
Created `PatchOptions`, `PatchOperation` structs for structured contract.
Implemented like Java SDK style, but added `SetCondition` method to support [Conditional Patch operation](https://learn.microsoft.com/en-us/rest/api/cosmos-db/patch-a-document#conditional-patch-operation)

File: sdk/data/azcosmos/cosmos_container.go
Comment (taiyokato) [reply sdk/data/azcosmos/cosmos_container.go]: See newly created `sdk/data/azcosmos/cosmos_item_request_patch_options.go`

File: sdk/data/azcosmos/cosmos_container.go
Comment (ealsur) [inline sdk/data/azcosmos/cosmos_container.go:371]: Based on the Azure Go guidelines, Options is used as a pattern for optional parameters (like ItemOptions at the end). 

One idea could be to change this to `patchOperations PatchOperations` (no pointer and renaming PatchOptions to PatchOperations).

@jhendrixMSFT do external contributions need to go through API review? How is that process?

File: sdk/data/azcosmos/cosmos_item_request_patch_options.go
Comment (ealsur) [inline sdk/data/azcosmos/cosmos_item_request_patch_options.go:45]: If the property is public, would we need a setter?

File: sdk/data/azcosmos/cosmos_item_request_patch_options.go
Comment (ealsur) [inline sdk/data/azcosmos/cosmos_item_request_patch_options.go:49]: If we are adding public methods to add operations, should `Operations` be public? And in that line of thought, should the consts like `PatchOperationTypeReplace` or the struct `PatchOperation` also be public?

Otherwise this has two venues, users can manually add PatchOperations to the Operations array or use these methods, if we want them to use the methods, then the other alternative should be removed by making them private?

File: sdk/data/azcosmos/README.md
Comment (ealsur) [inline sdk/data/azcosmos/README.md:164]: Once the public API is defined, we need to update these examples too

File: sdk/data/azcosmos/emulator_cosmos_item_test.go
Comment (ealsur) [inline sdk/data/azcosmos/emulator_cosmos_item_test.go:143]: Should we validate that the patch actually applied those modifications?

File: sdk/data/azcosmos/cosmos_item_request_patch_options.go
Comment (jhendrixMSFT) [reply sdk/data/azcosmos/cosmos_item_request_patch_options.go:45]: Agreed, this can be removed.

File: sdk/data/azcosmos/cosmos_item_request_patch_options.go
Comment (jhendrixMSFT) [inline sdk/data/azcosmos/cosmos_item_request_patch_options.go:36]: I think this can be removed as well.

File: sdk/data/azcosmos/cosmos_container.go
Comment (jhendrixMSFT) [reply sdk/data/azcosmos/cosmos_container.go:371]: Can you please create an APIView from these changes so we can review it?

CC @JeffreyRichter

Comment (ealsur) [reply sdk/data/azcosmos/cosmos_container.go:371]: @jhendrixMSFT The APIView is ready, I left some comments there, based on the outcome, we can finish polishing the PR and help @taiyokato complete it

File: sdk/data/azcosmos/cosmos_item_request_patch_options.go
Comment (ealsur) [inline sdk/data/azcosmos/cosmos_item_request_patch_options.go:9]: [From API review]  Make the enum private

File: sdk/data/azcosmos/cosmos_item_request_patch_options.go
Comment (ealsur) [inline sdk/data/azcosmos/cosmos_item_request_patch_options.go:23]: [From API review] Make private

File: sdk/data/azcosmos/cosmos_item_request_patch_options.go
Comment (ealsur) [inline sdk/data/azcosmos/cosmos_item_request_patch_options.go:30]: [From API review] Rename to `PatchOperations` make the fields private, they will be set through the helper functions

File: sdk/data/azcosmos/cosmos_item_request_patch_options.go
Comment (ealsur) [inline sdk/data/azcosmos/cosmos_item_request_patch_options.go:36]: [From API review] Remove, it's not needed

File: sdk/data/azcosmos/cosmos_item_request_patch_options.go
Comment (ealsur) [inline sdk/data/azcosmos/cosmos_item_request_patch_options.go:49]: [From API review] Rename to AppendReplace

File: sdk/data/azcosmos/cosmos_item_request_patch_options.go
Comment (ealsur) [inline sdk/data/azcosmos/cosmos_item_request_patch_options.go:58]: [From API review] Rename to AppendAdd

File: sdk/data/azcosmos/cosmos_item_request_patch_options.go
Comment (ealsur) [inline sdk/data/azcosmos/cosmos_item_request_patch_options.go:67]: [From API review] Rename to AppendSet

File: sdk/data/azcosmos/cosmos_item_request_patch_options.go
Comment (ealsur) [inline sdk/data/azcosmos/cosmos_item_request_patch_options.go:76]: [From API review] Rename to AppendRemove

File: sdk/data/azcosmos/cosmos_item_request_patch_options.go
Comment (ealsur) [inline sdk/data/azcosmos/cosmos_item_request_patch_options.go:84]: [From API review] Rename to AppendIncrement. Should the `value` be a number?

File: sdk/data/azcosmos/cosmos_container.go
Comment (ealsur) [inline sdk/data/azcosmos/cosmos_container.go:365]: [From API review]  These will be operations to probably the text would be: 

```suggestion
// ops - Operations to perform on the patch
```

**Review by ealsur — CHANGES_REQUESTED** [review]
@taiyokato I collected all the APIView comments into comments that start with `[From API review]`, let me know if they make sense. Once they are completed, we can do a final pass and merge

**Comment by ealsur** [comment]
@taiyokato let me know if you can make these changes or you need help, I will gladly chip in with commits on your fork

**Comment by ealsur** [comment]
@taiyokato do you want to continue this PR or should we take care of it?

**Comment by taiyokato** [comment]
@ealsur Sorry for the late response.
I saw there were some API reviews going on so thought a takeover might happen. 
I'm getting busy handling my personal life side and don't have much time working on stuff now.
If you could fork it and take over that'll be great. If not, I'll see when I can slowly work on this when I have the time.

**Comment by ealsur** [comment]
@taiyokato No worries, normally the API review just points at the required public API changes, but you can make them as author :)

I will take forward the PR with your contributions, focus on your personal life that is important, I hope everything is ok and thanks a lot for starting this work!

**Comment by ealsur** [comment]
Closing this PR in favor of https://github.com/Azure/azure-sdk-for-go/pull/19628 where we'll take the original contribution and apply the public API changes required

**Pull request closed without merging**
